Document Pod affinity namespaceSelector (#26820)
kubernetes/website: content/en/docs/concepts/policy/resource-quotas.md

| `BestEffort` | Match pods that have best effort quality of service. |
| `NotBestEffort` | Match pods that do not have best effort quality of service. |
| `PriorityClass` | Match pods that references the specified [priority class](/docs/concepts/configuration/pod-priority-preemption). |

> | `CrossNamespacePodAffinity` | Match pods that have cross-namespace pod [(anti)affinity terms](/docs/concepts/scheduling-eviction/assign-pod-node.md). |

Author: done

pods        0     10
```

> ### Cross Namespace Pod Affinity Quota

Reviewer: Style nit:
```suggestion
### Cross-namespace pod affinity quota
```

Author: done.

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

Operators can use `CrossNamespacePodAffinity` quota scope to limit which namespaces are allowed to have pods 

> with affinity terms that cross namespaces (i.e., allowed to set `namespaces` or `namespaceSelector` fields in pod affinity terms)

Reviewer: Is this quota scope only vaid if you have enabled the relevant feature gate?

Author: yes, I added a paragraph at the end.

> with affinity terms that cross namespaces (i.e., allowed to set `namespaces` or `namespaceSelector` fields in pod affinity terms)

Author: done.

> Preventing users from using cross-namespace affinity terms might be desired since a pod with anti-affinity constraints can block pods from all other namespaces 

Author: done

from getting scheduled in a failure domain. 

Using this scope operators can prevent certain namespaces (`foo-ns` in the example below) 
from having pods that use cross-namespace pod affinity by creating a resource quota object in
that namespace with `CrossNamespaceAffinity` scope and hard limit of 0:

```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: disable-cross-namespace-affinity
  namespace: foo-ns
spec:
  hard:
    pods: "0"
  scopeSelector:
    matchExpressions:
    - scopeName: CrossNamespaceAffinity
```

If operators want to disallow using `namespaces` and `namespaceSelector` by default, and 

> only allow it for specific namespaces,  they could configure `CrossNamespaceAffinity` 

Reviewer 2: Considering that this paragraph is in the "resource-quotas" page, mentioning "namespaces" and "namespaceSelector" in this context is confusing.

I'll take a look at the `ResourceQuotaConfiguration` and see if we can generate a reference API for it.

Author: sg, let me know if you have any suggestions.

pods        0     10
```

### Cross Namespace Pod Affinity Quota

{{< feature-state for_k8s_version="v1.21" state="alpha" >}}

Operators can use `CrossNamespacePodAffinity` quota scope to limit which namespaces are allowed to have pods 
with affinity terms that cross namespaces (i.e., allowed to set `namespaces` or `namespaceSelector` fields in pod affinity terms)

> Preventing users from using cross-namespace affinity terms might be necessary since a pod with anti-affinity constraint can block pods from all other namespaces 

Reviewer 3: ```suggestion
Preventing users from using cross-namespace affinity terms might be desired since a pod with anti-affinity constraints can block pods from all other namespaces 
```

from having pods that use cross-namespace pod affinity by creating a resource quota object in
that namespace with `CrossNamespaceAffinity` scope and hard limit of 0:

```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: disable-cross-namespace-affinity
  namespace: foo-ns
spec:
  hard:
    pods: "0"
  scopeSelector:
    matchExpressions:
    - scopeName: CrossNamespaceAffinity
```

If operators want to disallow using `namespaces` and `namespaceSelector` by default, and 
only allow it for specific namespaces,  they could configure `CrossNamespaceAffinity` 
as a limited resource by setting the kube-apiserver flag --admission-control-config-file
to the path of the following configuration file:

```yaml
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: "ResourceQuota"
  configuration:
    apiVersion: apiserver.config.k8s.io/v1
    kind: ResourceQuotaConfiguration
    limitedResources:
    - resource: pods
      matchScopes:
      - scopeName: CrossNamespaceAffinity
```

With the above configuration, pods can use `namespaces` and `namespaceSelector` in pod affinity only
if the namespace where they are created have a resource quota object with 

> `CrossNamespaceAffinity` scope and a hard limit equal to the number of pods that are
> allowed to.

Reviewer 3: ```suggestion
`CrossNamespaceAffinity` scope and a hard limit greater than or equal to the number of pods using those fields.
```

Author: done.

```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: disable-cross-namespace-affinity
  namespace: foo-ns
spec:
  hard:
    pods: "0"
  scopeSelector:
    matchExpressions:
    - scopeName: CrossNamespaceAffinity
```

If operators want to disallow using `namespaces` and `namespaceSelector` by default, and 
only allow it for specific namespaces, they could configure `CrossNamespaceAffinity` 
as a limited resource by setting the kube-apiserver flag --admission-control-config-file
to the path of the following configuration file:

```yaml
apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: "ResourceQuota"
  configuration:
    apiVersion: apiserver.config.k8s.io/v1
    kind: ResourceQuotaConfiguration
    limitedResources:
    - resource: pods
      matchScopes:
      - scopeName: CrossNamespaceAffinity
```

With the above configuration, pods can use `namespaces` and `namespaceSelector` in pod affinity only
if the namespace where they are created have a resource quota object with 
`CrossNamespaceAffinity` scope and a hard limit greater than or equal to the number of pods using those fields.

> This feature is alpha and disabled by default. You can enbale it by setting the

Reviewer 4: enable
